### PR TITLE
Change-Notification csharp sample.

### PR DESCRIPTION
**Pull request by Sudhir-MSFT**
Description
This sample shows a status of a user in personal scope  and demonstrates presence subscription  lifecycle followed by commands given to Bot

Language Used: C#

Proposed features
User Subscription, Bot, Status Change Notification, Bot SSO

Testing
After sucessfully installation of app you will get a sign in button. When sign in is complete then you get your current status in adapative card
image
image

# After that when the user status chagnes you will get notify about their status: 
- Change user status from available to busy like
image
- Change user status from busy to busy offline
 image

### CODE REVIEW

File: samples/change-notification/csharp/ChangeNotification/Bots/DialogBot.cs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+            await ConversationState.SaveChangesAsync(turnContext, false, cancellationToken);
+            await UserState.SaveChangesAsync(turnContext, false, cancellationToken);
+        }
+        protected override async Task OnMessageActivityAsync(ITurnContext<IMessageActivity> turnContext, CancellationToken cancellationToken)
+        {
+            Logger.LogInformation("Running dialog with Message Activity.");
+            await Dialog.RunAsync(turnContext, ConversationState.CreateProperty<DialogState>(nameof(DialogState)), cancellationToken);
+        }
+
+
```
Comment (Trinetra-MSFT) [inline samples/change-notification/csharp/ChangeNotification/Bots/DialogBot.cs:53]: correct space and format

File: samples/change-notification/csharp/ChangeNotification/Manifest/manifest.json [hunk 2]
```diff
@@ -0,0 +1,45 @@
+{
+  "$schema": "https://developer.microsoft.com/json-schemas/teams/v1.9/MicrosoftTeams.schema.json",
+  "manifestVersion": "1.9",
+  "version": "1.0.0",
+  "id": "<<App Id>>",
+  "packageName": "com.teams.questions.tracker",
```
Comment (Trinetra-MSFT) [inline samples/change-notification/csharp/ChangeNotification/Manifest/manifest.json:6]: Please rename this package to change notification

File: samples/change-notification/csharp/ChangeNotification/Manifest/manifest.json [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    "privacyUrl": "https://www.teams.com/privacy",
+    "termsOfUseUrl": "https://www.teams.com/termsofuser"
+  },
+  "icons": {
+    "color": "color.png",
+    "outline": "outline.png"
+  },
+  "name": {
+    "short": "ChangeNotificationApp",
+    "full": "Helps to track queries posted on developer community channels."
```
Comment (Trinetra-MSFT) [inline samples/change-notification/csharp/ChangeNotification/Manifest/manifest.json:19]: please correct full description, update what app does

File: samples/change-notification/csharp/ChangeNotification/Model/Notification.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        // The unique identifier for the webhooks subscription.
+        [JsonProperty(PropertyName = "subscriptionId")]
+        public string SubscriptionId { get; set; }
+
+        // Properties of the changed resource.
+        [JsonProperty(PropertyName = "resourceData")]
+        public ResourceData ResourceData { get; set; }
+
+        //[JsonProperty(PropertyName = "encryptedContent")]
+        //public Encryptedcontent EncryptedContent { get; set; }
```
Comment (Trinetra-MSFT) [inline samples/change-notification/csharp/ChangeNotification/Model/Notification.cs:42]: remove commented code

File: samples/change-notification/csharp/ChangeNotification/appsettings.json [hunk 5]
```diff
@@ -0,0 +1,7 @@
+{
+  "MicrosoftAppId": "",
+  "MicrosoftAppPassword": "",
+  "ConnectionName": "",
+  "BaseUrl": "",
+  "UserId": ""
```
Comment (Trinetra-MSFT) [inline samples/change-notification/csharp/ChangeNotification/appsettings.json:6]: Why we are putting `userId` in configuration?

File: samples/change-notification/csharp/README.md [hunk 6]
```diff
@@ -0,0 +1,89 @@
+﻿### Change Notification sample
+
+Bot Framework v4 ChangeNotification sample.
+
+This sample shows a status of a user in personal scope  and demonstrates presence subscription  lifecycle followed by commands given to Bot.
```
Comment (Trinetra-MSFT) [inline samples/change-notification/csharp/README.md:5]: `This sample app demonstrates sending change notifications to user presence in Teams based on user presence status`,

File: samples/change-notification/csharp/README.md [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+    ```bash
+    git clone https://github.com/OfficeDev/Microsoft-Teams-Samples.git
+    ```
+
+  A) Or from Visual Studio code
+
+  - Launch Visual Studio code
+  - File -> Open Folder
+  - Navigate to `samples/change-notification/csharp` folder
```
Comment (Trinetra-MSFT) [inline samples/change-notification/csharp/README.md:24]: Please rename this folder to `samples/graph-change-notifications/csharp`

**Review by Trinetra-MSFT — CHANGES_REQUESTED** [review]
Please update the root ReadMe file under graph section

File: samples/change-notification/csharp/ChangeNotification/appsettings.json (hunk 5 above)
Comment (Sudhir-MSFT) [reply samples/change-notification/csharp/ChangeNotification/appsettings.json:6]: It is required to subscribe the user presence on the basis of userId.

Comment (Trinetra-MSFT) [reply samples/change-notification/csharp/ChangeNotification/appsettings.json:6]: Can we read for from Graph API or bot context? I believe you are setting up the same user id in config who has installed this app

**Comment by Trinetra-MSFT** [comment]
@Sudhir-MSFT , Please update this sample link in the Root ReadMe file under graph section

File: samples/change-notification/csharp/ChangeNotification/appsettings.json (hunk 5 above)
Comment (Sudhir-MSFT) [reply samples/change-notification/csharp/ChangeNotification/appsettings.json:6]: Ya Trinetra, We can get it form bot context.

File: samples/change-notification/csharp/ChangeNotification/Model/Notification.cs (hunk 4 above)
Comment (Sudhir-MSFT) [reply samples/change-notification/csharp/ChangeNotification/Model/Notification.cs]: Done

**Review by Trinetra-MSFT — APPROVED** [review]

File: README.md [hunk 8]
```diff
@@ -185,6 +186,7 @@
 [graph#appcataloglifecycle#js]:samples/graph-appcatalog-lifecycle/nodejs
 
 [graph-activity-feed#cs]:samples/graph-activity-feed/csharp
+[graph-change-notification#cs]:samples/change-notification/csharp
```
Comment (Wajeed-msft) [inline README.md:189]: Could you please move this sample under **/samples/graph-change-notification/csharp**  folder?  Please change the root readme reference to point to correct location once folder is renamed.

File: samples/change-notification/csharp/ChangeNotification/Controllers/NotificationController.cs [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+namespace ChangeNotification.Controllers
+{
+    [Route("api/[controller]")]
+    [ApiController]
+    public class NotificationsController : Controller
+    {
+        private readonly IConfiguration _config;
+        private readonly ILogger _logger;
+        public static  ITurnContext stepContext;
```
Comment (Wajeed-msft) [inline samples/change-notification/csharp/ChangeNotification/Controllers/NotificationController.cs:26]: Storing stepContext is not correct way to send proactive message, please take a look at sample to [send proactive](https://github.com/microsoft/BotBuilder-Samples/blob/main/samples/csharp_dotnetcore/16.proactive-messages/Controllers/NotifyController.cs#L34) message. Please make the necessary changes.

File: samples/change-notification/csharp/ChangeNotification/Dialogs/MainDialog.cs [hunk 10]
```diff
@@ -0,0 +1,88 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using ChangeNotification.Helper;
```
Comment (Wajeed-msft) [inline samples/change-notification/csharp/ChangeNotification/Dialogs/MainDialog.cs:8]: Remove duplcate and unused usings form all the .cs files.

File: samples/change-notification/csharp/ChangeNotification/Helper/ChangeNotification.cs [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+using Microsoft.Bot.Schema;
+using AdaptiveCards;
+using Microsoft.Extensions.Caching.Memory;
+
+namespace ChangeNotification.Helper
+{
+    public class ChangeNotificationHelper
+    {
+
+        public Microsoft.Bot.Schema.Attachment ShowAdaptiveCard(string Header, ChangeNotification.Model.ResourceData InfoData)
```
Comment (Wajeed-msft) [inline samples/change-notification/csharp/ChangeNotification/Helper/ChangeNotification.cs:23]: Could you please rename this as this method just returns the Adpative card? Can we rename it to GetAvailabilityChangeCard()?

File: samples/change-notification/csharp/README.md [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+![image](https://user-images.githubusercontent.com/85864414/121880473-af1af780-cd2b-11eb-8166-837425ef186f.PNG)
+
+## Instruction for appsetting
+1. Provide appId and appPassword in the appsetting that is created in Azure.
+2. Provide UserId in appsetting for which you want to subscribe the user presence.
+3. Provide the ngrok url as  "BaseUrl" in appsetting on which application is running on.
+
+
+## Concepts introduced in this sample
+# After sucessfully installation of app you will get a sign in button. When sign in is complete then you get your current status in adapative card
```
Comment (Wajeed-msft) [inline samples/change-notification/csharp/README.md:71]: Could you please correct the formatting of these instructions? Currently it's showing with heading formattng. 
image

**Review by Wajeed-msft — CHANGES_REQUESTED** [review]

**Review by Wajeed-msft — APPROVED** [review]

**Pull request merged**